### DISCUSSION TITLE
Can't resolve 'fs'

### DISCUSSION THREAD

**Discussion by lualparedes** [vercel/next.js#72289] [Help] [answered]
### Summary

I'm trying to generate a static blog using Next 14, but I get a `Can't resolve 'fs'` error.

## Full error

![image](https://github.com/user-attachments/assets/60733a6a-b320-4874-bf7e-49a37846db30)

## Things I've tried

1. Using `getStaticProps` (same error)
2. Using `getServerSideProps` (same error)
3. Adding `config.resolve.fallback = { fs: false };` in `next.config.mjs` (`fs` gets imported but it's just an empty object

## Relevant code

`src/app/es/blog/[slug]/page.tsx`
```ts
import Post from "~/components/Post";
import { loadPosts } from "~/lib/blog";

import type { PostProps } from "~/types";

export async function generateStaticParams() {
  return loadPosts("es");
}

interface PostSpanishProps extends PostProps {}

const PostSpanish = ({ params: { slug } }: PostSpanishProps) => {
  return <Post slug={slug} />;
};

export default PostSpanish;
```

`src/lib/blog/index.ts`
```ts
import fs from "fs";
import path from "path";
import matter from "gray-matter";

import { assertPostsDirectoryExists } from "./assertions";
import {
  postsDirectoriesToRawPostStaticParams,
  rawPostStaticParamsToPostStaticParams,
} from "./mappers";

import type { Locale, PostStaticParams } from "~/types";
import type { RawPostStaticParams } from "./types";

export const getPostsDirectoryFor = (locale: Locale): string => {
  return path.join(path.resolve("content"), locale, "blog");
};

export const getPostsDirectories = (postsDirectory: string): string[] => {
  return fs.readdirSync(postsDirectory).filter((dir: string) => {
    return fs.statSync(path.join(postsDirectory, dir)).isDirectory();
  });
};

export const loadRawPostStaticParams = (
  locale: Locale,
): RawPostStaticParams[] => {
  const postsDirectory = getPostsDirectoryFor(locale);
  assertPostsDirectoryExists(postsDirectory);

  const rawPostStaticParams = getPostsDirectories(postsDirectory).map(
    postsDirectoriesToRawPostStaticParams(postsDirectory),
  );

  return rawPostStaticParams;
};

export const loadPosts = (locale: Locale = "en"): PostStaticParams[] => {
  return loadRawPostStaticParams(locale).map(
    rawPostStaticParamsToPostStaticParams,
  );
};
```

`src/lib/blog/assertions.ts`
```ts
import fs from "fs";

export const assertPostsDirectoryExists = (postsDirectory: string): void => {
  if (!fs.existsSync(postsDirectory)) {
    throw new Error(`Directory not found: ${postsDirectory}`);
  }
};

export const assertIndexMarkdownExists = (
  indexPath: string,
  dirPath: string,
): void => {
  if (!fs.existsSync(indexPath)) {
    throw new Error(`index.md not found in directory: ${dirPath}`);
  }
};
```

Any help is greatly appreciated! Thanks in advance! :blush: 

### Additional information

```js
Operating System:
  Platform: linux
  Arch: x64
  Version: #132~20.04.1-Ubuntu SMP Fri Aug 30 15:50:07 UTC 2024
  Available memory (MB): 15345
  Available CPU cores: 16
Binaries:
  Node: 20.15.1
  npm: 10.7.0
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.2.13 // An outdated version detected (latest is 15.0.2), upgrade is highly recommended!
  eslint-config-next: 14.2.5
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.5.3
Next.js Config:
  output: export
 ⚠ An outdated version detected (latest is 15.0.2), upgrade is highly recommended!
   Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
   Read more - https://nextjs.org/docs/messages/opening-an-issue
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

This often happens, because, one of the exports that's included in the files that import `fs`, is also being imported by a `use client` file, directly, or indirectly. Which means that Next.js is trying to pack it to be sent to the browser, but that's not gonna work.

You can try to quickly patch this by, adding this to your webpack key in `next.config.js`:

> [!WARNING]  
> Notice that we check the `nextRuntime` to be undefined, which targets the browser bundle

```js
// other stuff
  webpack: (config, { nextRuntime }) => {
    if (typeof nextRuntime === "undefined") {
      config.resolve.fallback = {
        ...config.resolve.fallback,
        fs: false,
      };
    }
   
    return config;
  },
// other stuff
```

You should though, try to fix this by re-organizing your imports, so that `fs` is not accidentally included into a client bundle.

**Reply by icyJoseph** [reply]
Also, are you using App Router or Pages Router?

You say: 

> Using getStaticProps (same error)
> Using getServerSideProps (same error)

Which are Pages Router, but also have: 

`src/app/es/blog/[slug]/page.tsx`, which is App Router

**Reply by icyJoseph** [reply]
Also updated my answer, with code that only blanks `fs` for client bundles.

**Reply by lualparedes** [reply]
@icyJoseph Thanks a lot! Your answer led me to find a component that was making use of a hook, which seemed to be the root cause of the issue because as soon as I removed the hook related code (including `"use client";`), the website finally got built without errors.

And for context regarding your question: I'm using the App Router, I tried with the other two "props getters" because they were solutions for similar issues I found online, and I was oblivious to the fact that they aren't supposed to be used with the App Router :sweat_smile:.
